File: internal/assets/paths.go
package assets

import (
	"fmt"
	"path/filepath"
	"strings"
)

const (
	// VaultLoc is the location of the vault.
	// This is the location of the documents and assets.
	VaultLoc = "internal/data/"
	// AssetsLoc is the location of the assets relative to the vault.
	AssetsLoc = "assets/"
	// PostsLoc is the location of the posts relative to the vault.
	PostsLoc = "posts/"
	// TagsLoc is the location of the tags relative to the vault.
	TagsLoc = "tags/"
	// ProjectsLoc is the location of the projects relative to the vault.
	ProjectsLoc = "projects/"
)

// Pathify returns the slugified path of a document or media asset.
func Pathify(s string) string {
	var path string
	var ok bool
	path, ok = strings.CutPrefix(s, AssetsLoc)
	if ok {
		return path
	}

	return strings.TrimSuffix(Slugify(s), filepath.Ext(s))
}

// Slugify returns the path to the document page or media asset page.
func Slugify(s string) string {
	var (
		path string
		ok   bool
	)
	s = strings.TrimSuffix(s, ".md")

	path, ok = strings.CutPrefix(s, PostsLoc)
	if ok {
		return path
	}
	path, ok = strings.CutPrefix(s, ProjectsLoc)
	if ok {
		return path
	}
	path, ok = strings.CutPrefix(s, TagsLoc)
	if ok {
		return path
	}
	path, ok = strings.CutPrefix(s, AssetsLoc)
	if ok {
		return path
	}

	panic(fmt.Errorf("failed to pathify %s", s))
}

// Static mapping of file extensions to content types to avoid reflection.
var contentTypes = map[string]string{
	".jpg":   "image/jpeg",
	".jpeg":  "image/jpeg",
	".png":   "image/png",
	".gif":   "image/gif",
	".svg":   "image/svg+xml",
	".webp":  "image/webp",
	".css":   "text/css",
	".txt":   "text/plain",
	".md":    "text/markdown",
	".pdf":   "application/pdf",
	".xml":   "application/xml",
	".zip":   "application/zip",
	".mp3":   "audio/mpeg",
	".mp4":   "video/mp4",
	".webm":  "video/webm",
	".wav":   "audio/wav",
	".mov":   "video/quicktime",
	".ico":   "image/x-icon",
	".woff":  "font/woff",
	".woff2": "font/woff2",
	".ttf":   "font/ttf",
	".otf":   "font/otf",
}

// GetContentType returns the content type for a file extension.
func GetContentType(path string) string {
	ext := filepath.Ext(path)
	if contentType, ok := contentTypes[ext]; ok {
		return contentType
	}
	// Fallback to standard library for unknown extensions
	return "application/octet-stream"
}

// BucketPath returns the path to the bucket for a given file path.
func BucketPath(path string) string {
	return "https://conneroisu.fly.storage.tigris.dev/assets/" + path
}

func isVideoType(contentType string) bool {
	switch contentType {
	case "video/mp4", "video/quicktime", "video/webm":
		return true
	default:
		return false
	}
}

File: live/src/models/paths.rs
use std::path::Path;

use super::{ASSETS_LOC, POSTS_LOC, PROJECTS_LOC, TAGS_LOC};

/// Slugify returns the slugified path of a document or media asset.
/// This removes the extension and directory prefix
pub fn slugify(path: &str) -> String {
    if let Some(stripped) = path.strip_prefix(ASSETS_LOC) {
        return stripped.to_string();
    }

    pathify(path)
        .trim_end_matches(['/', '\\'])
        .split('.')
        .next()
        .unwrap_or("")
        .to_string()
}

/// Pathify returns the path to the document page or media asset page by stripping prefixes
pub fn pathify(path: &str) -> String {
    if let Some(stripped) = path.strip_prefix(POSTS_LOC) {
        return stripped.to_string();
    }
    if let Some(stripped) = path.strip_prefix(PROJECTS_LOC) {
        return stripped.to_string();
    }
    if let Some(stripped) = path.strip_prefix(TAGS_LOC) {
        return stripped.to_string();
    }
    if let Some(stripped) = path.strip_prefix(ASSETS_LOC) {
        return stripped.to_string();
    }
    
    panic!("Failed to pathify {}", path);
}

/// Check if file extension indicates it's a valid media type
pub fn is_allowed_media_type(path: &str) -> bool {
    let allowed_extensions = [".jpg", ".jpeg", ".png", ".gif", ".webp", ".avif", ".tiff", ".svg", ".pdf"];
    
    if let Some(extension) = Path::new(path).extension() {
        if let Some(ext_str) = extension.to_str() {
            return allowed_extensions.contains(&format!(".{}", ext_str).as_str());
        }
    }
    
    false
}

/// Check if file extension indicates it's a valid document type
pub fn is_allowed_document_type(path: &str) -> bool {
    path.ends_with(".md")
}

/// Check if the file is either a media type or document type
pub fn is_allowed_asset(path: &str) -> bool {
    is_allowed_media_type(path) || is_allowed_document_type(path)
}

/// Get the content type for a file based on its extension
pub fn get_content_type(path: &str) -> String {
    match mime_guess::from_path(path).first_raw() {
        Some(mime) => mime.to_string(),
        None => "application/octet-stream".to_string(),
    }
}